GetBucketWebsite Details - NotImplemented error with third party S3 implementation

We are using netapp S3 and automate bucket creation and support using terraform. Bucket gets created but terraform fails due to 

```Error: error getting S3 Bucket website configuration: XNotImplemented: The request you provided implies functionality that is not implemented.```

### **Affected Resource(s)**

- aws_s3_bucket

### Terraform tf file

```provider "aws" {
  access_key                  = "mockaccess"
  region                      = "us-east-1"
  secret_key                  = "mockkey"
  s3_force_path_style         = true
  skip_credentials_validation = true
  skip_metadata_api_check     = true
  skip_requesting_account_id  = true
  skip_get_ec2_platforms = true
  endpoints {
     s3 = "https://xxxxxxxx:18082"
  }
}

resource "aws_s3_bucket" "b" {
  bucket = "tftest-bucket"
  acl="private"
}
```

### Debug output:
```
2020-08-13T17:31:53.307-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: 2020/08/13 17:31:53 [DEBUG] [aws-sdk-go] DEBUG: Request s3/GetBucketWebsite Details:
2020-08-13T17:31:53.307-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: ---[ REQUEST POST-SIGN ]-----------------------------
2020-08-13T17:31:53.307-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: GET /tftest-bucket?website= HTTP/1.1
2020-08-13T17:31:53.307-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: Host: xxxxxxxxxxxxxxxxxxxxxxxxxxxxx
2020-08-13T17:31:53.307-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: User-Agent: aws-sdk-go/1.32.12 (go1.13.7; linux; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.13.0 (+https://www.terraform.io)
2020-08-13T17:31:53.308-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: Authorization: AWS4-HMAC-SHA256 Credential=xxxxxxxxxxx/20200813/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=50295faba214b882a9c48010626cf2683cd3b6a9c9bee89023ddbd6ba71ba7ee
2020-08-13T17:31:53.308-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020-08-13T17:31:53.308-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: X-Amz-Date: 20200813T213153Z
2020-08-13T17:31:53.308-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: Accept-Encoding: gzip
2020-08-13T17:31:53.308-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4:
2020-08-13T17:31:53.308-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4:
2020-08-13T17:31:53.308-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: -----------------------------------------------------
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: 2020/08/13 17:31:53 [DEBUG] [aws-sdk-go] DEBUG: Response s3/GetBucketWebsite Details:
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: ---[ RESPONSE ]--------------------------------------
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: HTTP/1.1 501 Not Implemented
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: Connection: close
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: Content-Length: 263
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: Content-Type: application/xml
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: Date: Thu, 13 Aug 2020 21:31:53 GMT
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: Server: StorageGRID/11.3.0.3
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: X-Amz-Id-2: 12725169
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: X-Amz-Request-Id: 1597354313310942
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4:
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4:
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: -----------------------------------------------------
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: 2020/08/13 17:31:53 [DEBUG] [aws-sdk-go] <?xml version="1.0" encoding="UTF-8"?>
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: <Error><Code>XNotImplemented</Code><Message>The request you provided implies functionality that is not implemented.</Message><Resource>/tftest-bucket?website</Resource><RequestId>1597354313310942</RequestId></Error>
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4: 2020/08/13 17:31:53 [DEBUG] [aws-sdk-go] DEBUG: Validate Response s3/GetBucketWebsite failed, attempt 0/25, error XNotImplemented: The request you provided implies functionality that is not implemented.
2020-08-13T17:31:53.324-0400 [DEBUG] plugin.terraform-provider-aws_v2.70.0_x4:  status code: 501, request id: 1597354313310942, host id: 12725169
2020/08/13 17:31:53 [ERROR] eval: *terraform.EvalRefresh, err: error getting S3 Bucket website configuration: XNotImplemented: The request you provided implies functionality that is not implemented.
        status code: 501, request id: 1597354313310942, host id: 12725169
2020/08/13 17:31:53 [ERROR] eval: *terraform.EvalSequence, err: error getting S3 Bucket website configuration: XNotImplemented: The request you provided implies functionality that is not implemented.
        status code: 501, request id: 1597354313310942, host id: 12725169
2020/08/13 17:31:53 [TRACE] [walkRefresh] Exiting eval tree: aws_s3_bucket.b
2020/08/13 17:31:53 [TRACE] vertex "aws_s3_bucket.b": visit complete
2020/08/13 17:31:53 [TRACE] vertex "aws_s3_bucket.b": dynamic subgraph encountered errors
2020/08/13 17:31:53 [TRACE] vertex "aws_s3_bucket.b": visit complete
2020/08/13 17:31:53 [TRACE] vertex "aws_s3_bucket.b (expand)": dynamic subgraph encountered errors
2020/08/13 17:31:53 [TRACE] vertex "aws_s3_bucket.b (expand)": visit complete
2020/08/13 17:31:53 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/aws\"] (close)" errored, so skipping
2020/08/13 17:31:53 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020/08/13 17:31:53 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info

Error: error getting S3 Bucket website configuration: XNotImplemented: The request you provided implies functionality that is not implemented.
        status code: 501, request id: 1597354313310942, host id: 12725169
```

### Steps to Reproduce
Can be reproduced on custom endpoint implementation of s3.

### Proposed Solution
Add a flag which can skip features that are not implemented on this third party s3 providers.

--- a/.changelog/23278.txt
+++ b/.changelog/23278.txt
@@ -1,0 +1,3 @@
+```release-note:enhancement
+resource/aws_s3_bucket: Add error handling for `NotImplemented` and `XNotImplemented` errors when reading `acceleration_status`, `policy`, `request_payer`, or `website` into terraform state.
+```

--- a/.changelog/23278.txt
+++ b/.changelog/23278.txt
@@ -1,3 +1,7 @@
 ```release-note:enhancement
-resource/aws_s3_bucket: Add error handling for `NotImplemented` and `XNotImplemented` errors when reading `acceleration_status`, `policy`, `request_payer`, or `website` into terraform state.
+resource/aws_s3_bucket: Add error handling for `NotImplemented` errors when reading `acceleration_status`, `policy`, or `request_payer` into terraform state.
+```
+
+```release-note:enhancement
+resource/aws_s3_bucket: Add error handling for `MethodNotAllowed` and `XNotImplemented` errors when reading `website` into terraform state.
 ```

--- a/internal/service/s3/bucket.go
+++ b/internal/service/s3/bucket.go
@@ -909,7 +909,12 @@ func resourceBucketRead(d *schema.ResourceData, meta interface{}) error {
 		return nil
 	}
 
-	if err != nil && !tfawserr.ErrCodeEquals(err, ErrCodeNotImplemented, ErrCodeNoSuchWebsiteConfiguration, ErrCodeXNotImplemented) {
+	if err != nil && !tfawserr.ErrCodeEquals(err,
+		ErrCodeMethodNotAllowed,
+		ErrCodeNotImplemented,
+		ErrCodeNoSuchWebsiteConfiguration,
+		ErrCodeXNotImplemented,
+	) {
 		return fmt.Errorf("error getting S3 Bucket website configuration: %w", err)
 	}
 

--- a/internal/service/s3/bucket.go
+++ b/internal/service/s3/bucket.go
@@ -815,7 +815,7 @@ func resourceBucketRead(d *schema.ResourceData, meta interface{}) error {
 		return nil
 	}
 
-	if err != nil && !tfawserr.ErrCodeEquals(err, ErrCodeNoSuchBucketPolicy) {
+	if err != nil && !tfawserr.ErrCodeEquals(err, ErrCodeNoSuchBucketPolicy, ErrCodeNotImplemented) {
 		return fmt.Errorf("error getting S3 bucket (%s) policy: %w", d.Id(), err)
 	}
 
@@ -970,7 +970,7 @@ func resourceBucketRead(d *schema.ResourceData, meta interface{}) error {
 	}
 
 	// Amazon S3 Transfer Acceleration might not be supported in the region
-	if err != nil && !tfawserr.ErrCodeEquals(err, ErrCodeMethodNotAllowed, ErrCodeUnsupportedArgument) {
+	if err != nil && !tfawserr.ErrCodeEquals(err, ErrCodeMethodNotAllowed, ErrCodeUnsupportedArgument, ErrCodeNotImplemented) {
 		return fmt.Errorf("error getting S3 Bucket acceleration configuration: %w", err)
 	}
 
@@ -995,7 +995,7 @@ func resourceBucketRead(d *schema.ResourceData, meta interface{}) error {
 		return nil
 	}
 
-	if err != nil {
+	if err != nil && !tfawserr.ErrCodeEquals(err, ErrCodeNotImplemented) {
 		return fmt.Errorf("error getting S3 Bucket request payment: %s", err)
 	}
 

--- a/internal/service/s3/bucket.go
+++ b/internal/service/s3/bucket.go
@@ -909,7 +909,7 @@ func resourceBucketRead(d *schema.ResourceData, meta interface{}) error {
 		return nil
 	}
 
-	if err != nil && !tfawserr.ErrCodeEquals(err, ErrCodeNotImplemented, ErrCodeNoSuchWebsiteConfiguration) {
+	if err != nil && !tfawserr.ErrCodeEquals(err, ErrCodeNotImplemented, ErrCodeNoSuchWebsiteConfiguration, ErrCodeXNotImplemented) {
 		return fmt.Errorf("error getting S3 Bucket website configuration: %w", err)
 	}
 

--- a/internal/service/s3/errors.go
+++ b/internal/service/s3/errors.go
@@ -18,4 +18,9 @@ const (
 	ErrCodeReplicationConfigurationNotFound          = "ReplicationConfigurationNotFoundError"
 	ErrCodeServerSideEncryptionConfigurationNotFound = "ServerSideEncryptionConfigurationNotFoundError"
 	ErrCodeUnsupportedArgument                       = "UnsupportedArgument"
+
+	// ErrCodeXNotImplemented is returned from Third Party S3 implementations
+	// and so far has been noticed with calls to GetBucketWebsite.
+	// Reference: https://github.com/hashicorp/terraform-provider-aws/issues/14645
+	ErrCodeXNotImplemented = "XNotImplemented"
 )

